// asl-validator/dist/src/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StateMachineErrorCode = void 0;
var StateMachineErrorCode;
(function (StateMachineErrorCode) {
    StateMachineErrorCode["BranchOutboundTransitionTarget"] = "BRANCH_OUTBOUND_TRANSITION_TARGET";
    StateMachineErrorCode["DuplicateStateNames"] = "DUPLICATE_STATE_NAMES";
    StateMachineErrorCode["DuplicateFieldName"] = "DUPLICATE_FIELD_NAME";
    StateMachineErrorCode["InvalidJsonPath"] = "INVALID_JSON_PATH";
    StateMachineErrorCode["MapOutboundTransitionTarget"] = "MAP_OUTBOUND_TRANSITION_TARGET";
    StateMachineErrorCode["MissingTerminalState"] = "MISSING_TERMINAL_STATE";
    StateMachineErrorCode["MissingTransitionTarget"] = "MISSING_TRANSITION_TARGET";
    StateMachineErrorCode["SchemaValidationFailed"] = "SCHEMA_VALIDATION_FAILED";
    StateMachineErrorCode["TerminalStateWithNextError"] = "TERMINAL_STATE_WITH_NEXT";
    StateMachineErrorCode["WaitDurationError"] = "WAIT_DURATION";
    StateMachineErrorCode["TaskTimeoutError"] = "TASK_TIMEOUT";
    StateMachineErrorCode["TaskHeartbeatError"] = "TASK_HEARTBEAT";
    StateMachineErrorCode["MapItemProcessorError"] = "MAP_ITEM_PROCESSOR";
    StateMachineErrorCode["MapItemSelectorError"] = "MAP_ITEM_SELECTOR";
    StateMachineErrorCode["MapItemBatcherError"] = "MAP_ITEM_BATCHER";
    StateMachineErrorCode["MapToleratedFailureError"] = "MAP_TOLERATED_FAILURE";
    StateMachineErrorCode["MapMaxConcurrencyError"] = "MAP_CONCURRENCY_ERROR";
    StateMachineErrorCode["MapItemReaderMaxItemsError"] = "MAP_ITEMREADER_MAXITEM";
    StateMachineErrorCode["FailCauseProperty"] = "FAIL_CAUSE_PROPERTY";
    StateMachineErrorCode["FailErrorProperty"] = "FAIL_ERROR_PROPERTY";
})(StateMachineErrorCode || (exports.StateMachineErrorCode = StateMachineErrorCode = {}));
